testing/e2e/testkit.sh

```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
+
+function create_git_cached_rule_for_small_repo() {
+  local commit_hash=$1
+  local cache_dir=${2:-"default-dir"}
+
+  # Trim consecutive // characters since Bazel would fail if such exists on a target path
+  local trimmed_path=$(echo ${WORKSPACE_DIR} | tr -s '/')
+
+  if [[ ${cache_dir} == "default-dir" ]]; then
+    cat >${WORKSPACE_DIR}/WORKSPACE <<EOF
+load("//rules/git:git_cached_repository.bzl", "git_cached_repository")
+git_cached_repository(
+    name = "small_repo",
+    remote_url = "${REPO_SMALL_DIR}",
+    commit = "${commit_hash}",
+    branch = "master",
+)
+EOF
+  else
+    cat >${WORKSPACE_DIR}/WORKSPACE <<EOF
+load("//rules/git:git_cached_repository.bzl", "git_cached_repository")
+git_cached_repository(
+    name = "small_repo",
+    remote_url = "${REPO_SMALL_DIR}",
+    commit = "${commit_hash}",
+    branch = "master",
+    cache_directory = "${cache_dir}",
+)
+EOF
+  fi
+}
+
+function create_sh_lib_ref_small_repo() {
+  target_name=quote_reader_lib
+  target_label="quotes:${target_name}"
+
+  mkdir -p quotes
+  cat >quotes/BUILD <<EOF
+package(default_visibility = ["//visibility:public"])
+sh_library(
+    name = "${target_name}",
+    data = ["@small_repo//quotes:movies"],
+)
+EOF
+
+  echo ${target_label}
+}
+
+function create_sh_binary_ref_small_repo() {
+  target_name=quote_reader
+  target_label="quotes:${target_name}"
+
+  mkdir -p quotes
+  cat >quotes/BUILD <<EOF
+package(default_visibility = ["//visibility:public"])
+sh_binary(
+    name = "${target_name}",
+    srcs = ["quote_reader.sh"],
+    data = ["@small_repo//quotes:movies"],
+)
+EOF
+
+  cat >quotes/quote_reader.sh <<EOF
+#!/bin/sh
+cat ../small_repo/quotes/the_wizard_of_oz.txt
+EOF
+  chmod +x quotes/quote_reader.sh
+
+  echo ${target_label}
+}
+
+function create_new_commit_on_small_repo() {
+  # Create a new commit on the repo in test
+  cwd=$(pwd)
+  cd ${REPO_SMALL_DIR}
+  echo "correct dorothy, we are in bazel world" >${REPO_SMALL_DIR}/quotes/the_wizard_of_oz.txt
+  /usr/bin/git add "quotes/the_wizard_of_oz.txt" >>${TEST_log}
+  /usr/bin/git commit -m "Updated a quote to generate a new commit" >>${TEST_log}
+  new_commit=$(/usr/bin/git log -n 1 --pretty=format:%H)
```

Comment:
seems like you made a driver to create new commits on remote repository.
Why not generating this repo from scratch and drop usage in binary `tgz` file?

Reply:
> Why not generating this repo from scratch and drop usage in binary tgz file?
I didn't quite understand your suggestion, you opt in for creating a new repository index on the fly appending a commit to it?

If this is the case, doesn't tests data should be immutable? this is why the initial repository (and its index) are added as a tgz file.

By exposing the way the repository-in-test is created and/or what its data contains, you potentially can break all the testing suites.